pages/api/replicache-pull.ts

```diff
@@ -1,51 +1,109 @@
 import * as t from "io-ts";
 import type { NextApiRequest, NextApiResponse } from "next";
-import { ExecuteStatementCommandOutput } from "@aws-sdk/client-rds-data";
+import { ExecuteStatementCommandOutput, Field } from "@aws-sdk/client-rds-data";
 import { transact } from "../../backend/rds";
-import { getLastMutationID } from "../../backend/data";
+import { getCookieVersion, getLastMutationID } from "../../backend/data";
 import { must } from "../../backend/decode";
 
 export default async (req: NextApiRequest, res: NextApiResponse) => {
   console.log(`Processing pull`, req.body);
 
   const pull = must(pullRequest.decode(req.body));
+  let cookie = pull.baseStateID == "" ? 0 : parseInt(pull.baseStateID);
 
   console.time(`Reading all Shapes...`);
   let entries;
   let lastMutationID = 0;
 
   await transact(async (executor) => {
-    entries = (await executor("SELECT * FROM Shape")) ?? [];
-    lastMutationID = await getLastMutationID(executor, pull.clientID);
+    [entries, lastMutationID, cookie] = await Promise.all([
+      executor("SELECT * FROM Shape WHERE Version > :version", {
+        version: { longValue: cookie },
+      }),
+      getLastMutationID(executor, pull.clientID),
+      getCookieVersion(executor),
+    ]);
   });
+  console.log({ lastMutationID });
   console.timeEnd(`Reading all Shapes...`);
 
   // Grump. Typescript seems to not understand that the argument to transact()
   // is guaranteed to have been called before transact() exits.
   entries = (entries as any) as ExecuteStatementCommandOutput;
 
+  const resp: PullResponse = {
```

Comment:
@arv and I were discussing what to do in repc during pull when there's no new patch but the cookie value changes: https://rocicorp.slack.com/archives/G013XFG80JC/p1614626864010600?thread_ts=1614625340.004900&cid=G013XFG80JC. The right thing to do is write a new commit, which unless we special case it which i'm not excited about, requires replaying pending mutations. This seems fine to me, just bringing it to your attention.

In the case where we have no new patch it seems like we might want to return a nop pullresponse to avoid writing a commit in repc. By definition there's no values in the db with version >cookie, so it is no more expensive to scan next time from the old cookie vs the new cookie. (But also if we use an index it doesn't matter.)

We currently don't have an explicit 'nop' signal so the way to do this would be to return an empty patch and the same cookie (stateid). But maybe we should have an explicit signal so we don't have to infer intent?